### ISSUE TITLE
33% Failure rate when extracting data table

### ISSUE THREAD

**Issue by Lightosk** [healingbrew/UAssetParser#13]
(Put it here, because it's actually another issue than "not sure which branch to use")
I have an 33% failure rate, when trying to extract a large database. It comes out of the LSerializer.cs in UAssetParser. The function FString is the problematic one.
```cs
        // If the length of bytes is zero, the string is null.
        // If the length is above zero (positive), the string is UTF8.
        // If the length is below zero (negative), the string is UTF16LE.
        public static string FString(BinaryReader reader)
        {
            var bytes = reader.ReadInt32();
            if (bytes == 0) return string.Empty;
            else if (bytes > 0) return Encoding.UTF8.GetString(reader.ReadBytes(bytes), 0, bytes - 1);
            else return Encoding.Unicode.GetString(reader.ReadBytes((0 - bytes) * 2), 0, (0 - bytes) * 2 - 2);
        }
```
Some debugging did show, if it fails, it has got an unnatural big int. Then it says,
```
Unhandled exception. System.ArgumentOutOfRangeException: Index and count must refer to a location within the buffer. (Parameter 'bytes')
   at System.Text.UTF8Encoding.GetString(Byte[] bytes, Int32 index, Int32 count)
   at UAssetParser.LSerializer.FString(BinaryReader reader) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\LSerializer.cs:line 153
   at UAssetParser.LSerializer.Deserialize(BinaryReader data, Type T, Nullable`1 countRef) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\LSerializer.cs:line 57
   at UAssetParser.LSerializer.DeserializeComplex(BinaryReader data, Type T, Nullable`1 countRef) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\LSerializer.cs:line 84
   at UAssetParser.LSerializer.Deserialize(BinaryReader data, Type T, Nullable`1 countRef) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\LSerializer.cs:line 57
   at UAssetParser.LSerializer.Deserialize[T](BinaryReader data) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\LSerializer.cs:line 51
   at UAssetParser.Objects.Visitors.Unreal.UTextProperty.PropertyVisitor(BinaryReader reader, FPropertyTag baseTag, FPackageFileSummary summary) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\Visitors\Unreal\UTextProperty.cs:line 33
   at UAssetParser.Objects.Visitors.VisitorFactory.Visit(BinaryReader reader, FPropertyTag baseTag, FPackageFileSummary summary) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\Visitors\VisitorFactory.cs:line 104
   at UAssetParser.Objects.UObject..ctor(BinaryReader data, FPackageFileSummary summary, Boolean pad, FObjectExport export) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\UObject.cs:line 37
   at UAssetParser.Objects.Visitors.Unreal.UStructProperty.PropertyVisitor(BinaryReader reader, FPropertyTag baseTag, FPackageFileSummary summary) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\Visitors\Unreal\UStructProperty.cs:line 29
   at UAssetParser.Objects.Visitors.VisitorFactory.Visit(BinaryReader reader, FPropertyTag baseTag, FPackageFileSummary summary) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\Visitors\VisitorFactory.cs:line 104
   at UAssetParser.Objects.UObject..ctor(BinaryReader data, FPackageFileSummary summary, Boolean pad, FObjectExport export) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\UObject.cs:line 37
   at UAssetParser.Objects.Visitors.Unreal.UDataTable.Deserialize(BinaryReader reader, FObjectExport export, FPackageFileSummary summary) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\Visitors\Unreal\UDataTable.cs:line 20
   at UAssetParser.Objects.Visitors.VisitorFactory.VisitSubtype(BinaryReader reader, FObjectExport export, FPackageFileSummary summary) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\Visitors\VisitorFactory.cs:line 130
   at UAssetParser.Objects.UObject..ctor(BinaryReader data, FPackageFileSummary summary, Boolean pad, FObjectExport export) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\Objects\UObject.cs:line 43
   at UAssetParser.UAsset..ctor(Stream uassetStream, Stream uexpStream, Stream ubulkStream, Boolean keepOpen) in C:\Users\Progosk\Downloads\UAssetParser-master\UAssetParser\UAsset.cs:line 34
   at UAsset2Json.Program.Export(String path) in C:\Users\Progosk\Downloads\UAssetParser-master\UAsset2Json\Program.cs:line 38
   at UAsset2Json.Program.ExportDir(String[] args) in C:\Users\Progosk\Downloads\UAssetParser-master\UAsset2Json\Program.cs:line 26
   at UAsset2Json.Program.Main(String[] args) in C:\Users\Progosk\Downloads\UAssetParser-master\UAsset2Json\Program.cs:line 14
```
The .uasset and .uexp files are attached. If needed. Comment if more information is needed.
[Problematic_UAsset_Files.zip](https://github.com/healingbrew/UAssetParser/files/4565857/Problematic_UAsset_Files.zip)

**Comment by Kein** [comment]
Okay, so it looks like that asset has non-generic StructProperty unique to that specific game called `YActorUIWidgetData` and therefore, instead of Struct, old parser creates generic UObject and tries to process contained TextProperty normally but it looks like it has no StreamOffset value when packed like that and I dont see any indication that would help to figure it out in the asset itself.

Basically as I understood properties packed there in a very game-specific way, which means for that game there is probably needs to be a custom condition.

**Comment by Kein** [comment]
Can you test this version on some more assets that have TextProperty?
[test.zip](https://github.com/healingbrew/UAssetParser/files/4568542/test.zip)

**Comment by Lightosk** [comment]
Are you really sure I should open this? I think it's better if I build it myself, so I can ensure I get no Malware. Sorry.
![grafik](https://user-images.githubusercontent.com/58272323/80943245-0b041480-8de7-11ea-840f-6ffe606c29b4.png)

**Comment by Kein** [comment]
Huh weird. Ignore it then, I was just interested if you have any assets with TextProperty packed into struct/array.

Either way `core` now can handle your assets.

**Comment by Lightosk** [comment]
Wahooo! Thanks for your help, I'll try it now.
